File: puzzles/php/shadows-knight1.php
<?php
fscanf(STDIN, "%d %d", $width, $height);
fscanf(STDIN, "%d", $nbTurns);
fscanf(STDIN, "%d %d", $x, $y);

// Indicates the search area coordinates
$minX = 0; // minimum x value
$maxX = $width - 1; // maximum x value
$minY = 0; // minimum y value
$maxY = $height - 1; // maximum y value

// game loop
while (true) {
  fscanf(STDIN, "%s", $bombDir);

  if ($bombDir == "U") { // UP
    $maxY = $y - 1;
    $y = ($minY + $maxY) / 2;
  } else if ($bombDir == "UR") { // UP-RIGHT
    $minX = $x + 1;
    $maxY = $y - 1;
    $x = ($minX + $maxX) / 2;
    $y = ($minY + $maxY) / 2;
  } else if ($bombDir == "R") { // RIGHT
    $minX = $x + 1;
    $x = ($minX + $maxX) / 2;
  } else if ($bombDir == "DR") { // DOWN-RIGHT
    $minX = $x + 1;
    $minY = $y + 1;
    $x = ($minX + $maxX) / 2;
    $y = ($minY + $maxY) / 2;
  } else if ($bombDir == "D") { // DOWN
    $minY = $y + 1;
    $y = ($minY + $maxY) / 2;
  } else if ($bombDir == "DL") { // DOWN-LEFT
    $maxX = $x - 1;
    $minY = $y + 1;
    $x = ($minX + $maxX) / 2;
    $y = ($minY + $maxY) / 2;
  } else if ($bombDir == "L") { // LEFT
    $maxX = $x - 1;
    $x = ($minX + $maxX) / 2;
  } else { // UL or UP-LEFT
    $maxX = $x - 1;
    $maxY = $y - 1;
    $x = ($minX + $maxX) / 2;
    $y = ($minY + $maxY) / 2;
  } // else

  $x = (int) $x;
  $y = (int) $y;

  // the location of the next window Batman should jump to.
  echo($x ." " . $y . "\n");
}
?>

File: puzzles/python3/shadows_knight1.py
if __name__ == "__main__":
    width, height = map(int,  input().split())
    nb_turns = int(input())
    x, y = map(int, input().split())

    min_x = 0
    max_x = width - 1
    min_y = 0
    max_y = height - 1

    # game loop
    while True:
        bomb_dir = input()

        if bomb_dir == "U":
            max_y = y - 1
            min_x = max_x = x
        elif bomb_dir == "UR":
            min_x = x + 1
            max_y = y - 1
        elif bomb_dir == "R":
            min_x = x + 1
            min_y = max_y = y
        elif bomb_dir == "DR":
            min_x = x + 1
            min_y = y + 1
        elif bomb_dir == "D":
            min_y = y + 1
            min_x = max_x = x
        elif bomb_dir == "DL":
            max_x = x - 1
            min_y = y + 1
        elif bomb_dir == "L":
            max_x = x - 1
            min_y = max_y = y
        elif bomb_dir == "UL":
            max_x = x - 1
            max_y = y - 1

        x = int((min_x + max_x) / 2)
        y = int((min_y + max_y) / 2)
        # the location of the next window Batman should jump to.
        print(f"{x} {y}")
